gtsfm/utils/metrics.py

```diff
@@ -451,3 +453,70 @@ def compute_percentage_change(x: float, y: float) -> float:
         percentage change (may be positive or negative).
     """
     return (y - x) / (x + EPSILON) * 100
+
+
+def compute_psnr_after_downsampling(
+    original_point_cloud: np.ndarray, downsampled_point_cloud: np.ndarray
+) -> GtsfmMetric:
+    """Compute PSNR between original point cloud and downsampled point cloud
+    A larger PSNR shows that the downsampled point cloud consists better with the original point cloud
+
+    Args:
+        original_point_cloud: original dense point cloud before downsampling
+        downsampled_point_cloud: dense point cloud after downsampling
+
+    Returns:
+        GtsfmMetric: PSNR between original point cloud and downsampled point cloud
+    """
+    diagnose_voxel_scale = np.linalg.norm(mvs_utils.estimate_voxel_scales(original_point_cloud))
+    original_tree = KDTree(data=original_point_cloud)
+    downsampled_tree = KDTree(data=downsampled_point_cloud)
+
+    d_downsampled_to_original, _ = original_tree.query(downsampled_point_cloud)
+    d_original_to_downsampled, _ = downsampled_tree.query(original_point_cloud)
+
+    def RMS(data: np.ndarray) -> float:
+        """Utility function to calculate the root mean square of the given data
+
+        Args:
+            data: input data for evaluation
+
+        Returns:
+            the root mean square of the given data
+        """
+        return np.sqrt(np.square(data).mean())
+
+    psnr = 20.0 * np.log10(diagnose_voxel_scale / max(RMS(d_downsampled_to_original), RMS(d_original_to_downsampled)))
+
+    return GtsfmMetric(name="downsampling PSNR", data=psnr)
+
+
+def get_voxel_downsampling_metrics(
+    min_voxel_size: float, original_point_cloud: np.ndarray, downsampled_point_cloud: np.ndarray
+) -> GtsfmMetricsGroup:
+    """Collect and compute metrics for voxel downsampling
+    Args:
+        min_voxel_size: minimum voxel size for voxel downsampling
+        original_point_cloud: original dense point cloud before downsampling
+        downsampled_point_cloud: dense point cloud after downsampling
+
+    Returns:
+        GtsfmMetricsGroup: voxel downsamping metrics group
+    """
+    downsampling_metrics = []
+    downsampling_metrics.append(GtsfmMetric(name="voxel size for downsampling", data=min_voxel_size))
+    downsampling_metrics.append(
+        GtsfmMetric(name="point cloud size before downsampling", data=original_point_cloud.shape[0])
+    )
+    downsampling_metrics.append(
+        GtsfmMetric(name="point cloud size after downsampling", data=downsampled_point_cloud.shape[0])
+    )
+    downsampling_metrics.append(
+        GtsfmMetric(name="compression ratio", data=original_point_cloud.shape[0] / downsampled_point_cloud.shape[0])
+    )
+    downsampling_metrics.append(
+        compute_psnr_after_downsampling(
```

Comment:
could you return the float psnr from  compute_psnr_after_downsampling , and then create the GtsfmMetric here like you do for other metrics?